pkg/apis/tenancy/v1alpha1/types_workspacetype.go

```diff
@@ -0,0 +1,267 @@
+// composed together to add functionality to the owning WorkspaceType.
+type WorkspaceTypesExtension struct {
+	// with are WorkspaceTypes whose initializers are added to the list
+	// for the owning type, and for whom the owning type becomes an alias, as long
+	// as all of their required types are not mentioned in without.
+	//
+	// +optional
+	With []WorkspaceTypesReference `json:"with,omitempty"`
+}
+
+// These are valid conditions of WorkspaceType.
+const (
+	WorkspaceTypesVirtualWorkspaceURLsReady conditionsv1alpha1.ConditionType = "VirtualWorkspaceURLsReady"
+	ErrorGeneratingURLsReason                                                = "ErrorGeneratingURLs"
+)
+
+// WorkspaceTypeStatus defines the observed state of WorkspaceType.
+type WorkspaceTypeStatus struct {
+	// conditions is a list of conditions that apply to the APIExport.
+	//
+	// +optional
+	Conditions conditionsv1alpha1.Conditions `json:"conditions,omitempty"`
+
+	// virtualWorkspaces contains all APIExport virtual workspace URLs.
+	// +optional
+	VirtualWorkspaces []VirtualWorkspace `json:"virtualWorkspaces,omitempty"`
+}
+
+type VirtualWorkspace struct {
+	// url is a WorkspaceType initialization virtual workspace URL.
+	//
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:format:URL
+	// +required
+	URL string `json:"url"`
+}
+
+func (in *WorkspaceType) GetConditions() conditionsv1alpha1.Conditions {
+	return in.Status.Conditions
+}
+
+func (in *WorkspaceType) SetConditions(conditions conditionsv1alpha1.Conditions) {
+	in.Status.Conditions = conditions
+}
+
+// WorkspaceTypeList is a list of cluster workspace types
+//
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type WorkspaceTypeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []WorkspaceType `json:"items"`
+}
+
+// WorkspaceAPIBindingsInitializer is a special-case initializer that waits for APIBindings defined
+// on a WorkspaceType to be created.
+const WorkspaceAPIBindingsInitializer corev1alpha1.LogicalClusterInitializer = "system:apibindings"
+
+const (
+	// WorkspacePhaseLabel holds the ClusterWorkspace.Status.Phase value, and is enforced to match
+	// by a mutating admission webhook.
+	WorkspacePhaseLabel = "tenancy.kcp.dev/phase"
+	// WorkspaceInitializerLabelPrefix is the prefix for labels which match Workspace.Status.Initializers,
+	// and the set of labels with this prefix is enforced to match the set of initializers by a mutating admission
+	// webhook.
+	WorkspaceInitializerLabelPrefix = "initializer.internal.kcp.dev/"
+)
+
+const (
+	// RootWorkspaceTypeName is a reference to the root logical cluster, which has no cluster workspace type
+	RootWorkspaceTypeName = WorkspaceTypesName("root")
+)
+
+var (
+	// RootWorkspaceTypeReference is a reference to the root logical cluster, which has no cluster workspace type
+	RootWorkspaceTypeReference = WorkspaceTypesReference{
```

Comment:
This is mutable by package consumers - do we want to protect against that?

Reply:
maybe. This is just moved code.

Reply:
even better: remove this over abstraction. This type is not special.